oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/LogoutHandler.java

```diff
@@ -231,45 +168,133 @@ private APIGatewayProxyResponseEvent validateClientIDAgainstClientRegistry(
                             } else {
                                 LOG.info(
                                         "The post_logout_redirect_uri is present in logout request and client registry. Value is {}",
                                         uri);
-                                cloudwatchMetricsService.incrementLogout(Optional.of(clientID));
-                                return logoutService.generateLogoutResponse(
-                                        URI.create(uri),
-                                        state,
-                                        Optional.empty(),
-                                        input,
-                                        Optional.of(clientID),
-                                        sessionId);
+                                return true;
                             }
                         })
-                .orElseGet(
-                        () -> {
-                            LOG.info(
-                                    "post_logout_redirect_uri is NOT present in logout request. Generating default logout response");
-                            return logoutService.generateDefaultLogoutResponse(
-                                    state, input, Optional.of(clientID), sessionId);
-                        });
+                .orElseGet(() -> false);
+    }
+
+    private APIGatewayProxyResponseEvent logout(
+            String idTokenHint,
+            Session session,
+            String clientID,
+            String uri,
+            Optional<String> state,
+            APIGatewayProxyRequestEvent input) {
+        CookieHelper.SessionCookieIds sessionCookieIds =
+                cookieHelper.parseSessionCookie(input.getHeaders()).orElseThrow();
+
+        attachSessionIdToLogs(session);
+        attachLogFieldToLogs(CLIENT_SESSION_ID, sessionCookieIds.getClientSessionId());
+        attachLogFieldToLogs(GOVUK_SIGNIN_JOURNEY_ID, sessionCookieIds.getClientSessionId());
+
+        if (!session.getClientSessions().contains(sessionCookieIds.getClientSessionId())) {
+            LOG.warn("Client Session ID does not exist");
+            return logoutService.generateErrorLogoutResponse(
+                    Optional.empty(),
+                    new ErrorObject(OAuth2Error.INVALID_REQUEST_CODE, "invalid session"),
+                    input,
+                    Optional.empty(),
+                    Optional.of(session.getSessionId()));
+        }
+
+        if (!doesIDTokenExistInSession(idTokenHint, session)) {
+            LOG.warn("ID token does not exist");
+            return logoutService.generateErrorLogoutResponse(
+                    Optional.empty(),
+                    new ErrorObject(
+                            OAuth2Error.INVALID_REQUEST_CODE, "unable to validate id_token_hint"),
+                    input,
+                    Optional.empty(),
+                    Optional.of(session.getSessionId()));
+        }
```

Comment:
Have you worked this through in detail? I'm a little worried some of this runs counter to the point of the ticket.

Reply:
What if, for example, they have logged out and back in with service B in another tab while using service A? When service A sends the original token hint this will fail I think